iree/compiler/Dialect/HAL/Transforms/InlineDeviceSwitches.cpp

```diff
@@ -76,62 +77,82 @@ static Value buildConditionExpression(Location loc, Value device,
   return {};
 }
 
-// Outlines a condition region from a switch op into a standalone function.
-static FuncOp outlineConditionRegion(StringRef funcName,
-                                     Region &conditionRegion,
-                                     ArrayRef<Type> resultTypes,
-                                     OpBuilder moduleBuilder) {
+// Inlines a condition region from a switch op into the function at the given
+// point. This assumes that the insertion point will only be reached if the
+// condition the region is predicated on is true.
+static void inlineConditionRegion(OperandRange regionArgs,
+                                  Region &conditionRegion, Block *exitBlock,
+                                  OpBuilder funcBuilder) {
+  assert(!conditionRegion.empty() && "source regions must not be empty");
+
+  // Split the insertion block.
+  auto *insertBlock = funcBuilder.getBlock();
+  auto *postInsertBlock = &*std::next(insertBlock->getIterator());
+
+  // Remap arguments from the function values captured by the switch into the
+  // entry block arguments for the region.
   auto &entryBlock = conditionRegion.front();
-  auto funcType = moduleBuilder.getFunctionType(
-      llvm::to_vector<4>(
-          llvm::map_range(entryBlock.getArguments(),
-                          [](BlockArgument arg) { return arg.getType(); })),
-      resultTypes);
-  auto funcOp = moduleBuilder.create<FuncOp>(
-      conditionRegion.getLoc(), funcName, funcType, ArrayRef<NamedAttribute>{});
-  funcOp.getBody().takeBody(conditionRegion);
-  SymbolTable::setSymbolVisibility(funcOp, SymbolTable::Visibility::Private);
-
-  // Replace hal.return statements with normal std.return. This ensures that
-  // normal matchers/inlining/etc works as we continue transformation.
-  for (auto &block : funcOp.getBlocks()) {
-    if (auto returnOp = dyn_cast<IREE::HAL::ReturnOp>(block.back())) {
-      OpBuilder builder(returnOp);
-      builder.create<mlir::ReturnOp>(
-          returnOp.getLoc(), llvm::to_vector<4>(returnOp.getOperands()));
+  assert(regionArgs.size() == entryBlock.getNumArguments() &&
+         "switch capture args must match region args");
+  BlockAndValueMapping mapping;
+  for (auto argPair : llvm::zip(regionArgs, entryBlock.getArguments())) {
+    auto outerValue = std::get<0>(argPair);
+    auto innerValue = std::get<1>(argPair);
+    assert(outerValue.getType() == innerValue.getType() &&
+           "capture arg types must match");
+    innerValue.replaceAllUsesWith(outerValue);
+  }
+
+  // Splice in the region blocks.
+  auto *insertRegion = insertBlock->getParent();
+  insertRegion->getBlocks().splice(
+      postInsertBlock->getIterator(), conditionRegion.getBlocks(),
```

Comment:
So I guess splice takes the place you're splicing before not the place you're splicing after? There isn't by any chance any documentation for splice anywhere, is there? The only thing I found is "// Splice members - defined in terms of transfer...", which leaves much to be desired :stuck_out_tongue: 

Can you use cloneInto? I think that gives a higher-level API

Reply:
I am explicitly not using cloneInto, as that *clones* and this *splices* in-place. The same reason the InlinerUtils uses it.